pandas/tests/test_groupby.py

```diff
@@ -720,6 +720,38 @@ def test_agg_period_index(self):
         grouped = df.groupby(df.index.month)
         list(grouped)
 
+    def test_agg_dict_parameter_cast_result_dtypes(self):
+        # GH 12821
+
+        df = DataFrame(
+            {'class': ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D'],
+             'time': date_range('1/1/2011', periods=8, freq='H')})
+        df.loc[[0, 1, 2, 5], 'time'] = None
+
+        # test for `first` function
+        exp = df.loc[[0, 3, 4, 6]].set_index('class')
+
+        res = df.groupby('class').first()
```

Comment:
need to test

``df.groupby('class').time.first()`` and so on (e.g. repeat with ``.agg`` both forms as well)

if making a loop helps to simplify testing, pls do so